File: golang/_0018_4Sum/main.go
package main

import "sort"

func fourSum(nums []int, target int) [][]int {
	n, res := len(nums), make([][]int, 0)
	if n >= 4 {
		sort.Ints(nums)
		for i := 0; i < n-3; i++ {
			if i > 0 && nums[i] == nums[i-1] {
				continue
			}

			for j := i + 1; j < n-2; j++ {
				if j > i+1 && nums[j] == nums[j-1] {
					continue
				}

				for m, n := j+1, n-1; m < n; {
					sum := nums[i] + nums[j] + nums[m] + nums[n]
					if sum == target {
						res = append(res, []int{nums[i], nums[j], nums[m], nums[n]})
						m++
						for m < n && nums[m] == nums[m-1] {
							m++
						}
					} else if sum < target {
						m++
					} else {
						n--
					}
				}
			}
		}
	}

	return res
}

File: python3/_0018_4Sum/main.py
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if not nums:
            return []

        res, n = [], len(nums)
        nums.sort()

        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, n):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue

                u, v = j + 1, n - 1
                while u < v:
                    t = nums[i] + nums[j] + nums[u] + nums[v]
                    if t == target:
                        res.append([nums[i], nums[j], nums[u], nums[v]])
                        u += 1
                        while u < v and nums[u] == nums[u - 1]:
                            u += 1
                    elif t > target:
                        v -= 1
                    else:
                        u += 1

        return res
